### ISSUE TITLE
Dropout makes LSTM non-deterministic

### ISSUE THREAD

**Issue by pozimekSG** [pytorch/pytorch#90168]
### 🐛 Describe the bug

Adding dropout to the LSTM module makes it behave in a non-deterministic way. The documentation gave me an impression of the module behaving in a deterministic fashion if the correct procedure is followed.

```
#lstm_test.py
import torch
import torch.nn as nn

torch.use_deterministic_algorithms(True)
seed = 1

for dropout in [0.5, 0]:
    model = nn.LSTM(4, 4, 2, dropout=dropout).cuda()
    model.train()

    for i in range(3):
        torch.manual_seed(seed)
        torch.cuda.manual_seed(seed)

        data = torch.randn(4, 4)
        if i > 0:
            print(torch.equal(data, pre_data))
        pre_data = data
        data = data.cuda()
        out, _ = model(data)
        loss = out.sum()
        
        print(f"D{dropout}, {i}, Loss: {loss.item()}")
```

Running `CUBLAS_WORKSPACE_CONFIG=:4096:8 python lstm_test.py` produces the following output:

D0.5, 0, Loss: 0.8866389989852905
True
D0.5, 1, Loss: 0.4514765739440918
True
D0.5, 2, Loss: 0.7463627457618713
D0, 0, Loss: -1.6037826538085938
True
D0, 1, Loss: -1.6037826538085938
True
D0, 2, Loss: -1.6037826538085938

I was expecting the loss to be identical between the first three runs with dropout=0.5 just as it is the same between the second three runs. This issue is making it impossible for me to reproduce a NaN gradient bug. Rerunning with `CUBLAS_WORKSPACE_CONFIG=:16:8` and `CUBLAS_WORKSPACE_CONFIG=:4096:2` does not resolve the issue.

### Versions

Collecting environment information...
PyTorch version: 1.12.1
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2
Libc version: glibc-2.27

Python version: 3.9.13 | packaged by conda-forge | (main, May 27 2022, 16:56:21)  [GCC 10.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-128-generic-x86_64-with-glibc2.27
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: 
GPU models and configuration: 
GPU 0: NVIDIA TITAN RTX
GPU 1: NVIDIA TITAN RTX

Nvidia driver version: 470.141.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.3
[pip3] pytorch-model-summary==0.1.1
[pip3] torch==1.12.1
[pip3] torch-tb-profiler==0.4.0
[pip3] torchaudio==0.12.1
[pip3] torchvision==0.13.1
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.3.1               h2bc3f7f_2  
[conda] libblas                   3.9.0            14_linux64_mkl    fastchan
[conda] libcblas                  3.9.0            14_linux64_mkl    fastchan
[conda] liblapack                 3.9.0            14_linux64_mkl    fastchan
[conda] mkl                       2022.1.0           hc2b9512_224  
[conda] numpy                     1.23.3           py39hba7629e_0    fastchan
[conda] pytorch                   1.12.1          py3.9_cuda11.3_cudnn8.3.2_0    pytorch
[conda] pytorch-model-summary     0.1.1                      py_0    conda-forge
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torch-tb-profiler         0.4.0                    pypi_0    pypi
[conda] torchaudio                0.12.1               py39_cu113    pytorch
[conda] torchvision               0.13.1               py39_cu113    pytorch

**Comment by ptrblck** [comment]
It seems the usage of cuDNN's LSTM is not deterministic at the moment.
You could disable it for now as a workaround and set the corresponding cublas env variables to make it deterministic again via `torch.backends.cudnn.enabled = False`.
